daffodil-runtime1/src/main/scala/edu/illinois/ncsa/daffodil/processors/parsers/Parser.scala

```diff
@@ -189,76 +193,107 @@ class AltCompParser(context: RuntimeData, val childParsers: Seq[Parser])
+          //
+          if (pstate.discriminator == true) {
+            log(LogLevel.Debug, "Failure, but discriminator true. Additional alternatives discarded.")
+            // If so, then we don't run the next alternative, we
+            // consume this discriminator status result (so it doesn't ripple upward)
+            // and return the failed state with all the diagnostics.
+            //
+            val allDiags = new AltParseFailed(context.schemaFileLocation, pstate, diagnostics.reverse)
+            pstate.discard(pBefore) // because disc set, we don't unwind side effects on input stream & infoset
+            pBefore = null
+            pstate.setFailed(allDiags)
+            returnFlag = true
+          } else {
+            //
+            // Here we have a failure, but no discriminator was set, so we try the next alternative.
+            // Which means we just go around the loop
+            //
+            // But we have to unwind side-effects on input-stream, infoset, variables, etc.
+            pstate.reset(pBefore)
+            pBefore = null
+            returnFlag = false
+          }
         }
       }
-    }
-    Assert.invariant(i <= limit)
-    if (returnFlag == false) {
-      Assert.invariant(i == limit)
-      // Out of alternatives. All of them failed.
-      val allDiags = new AltParseFailed(context.schemaFileLocation, pstate, diagnostics.reverse)
-      pstate.setFailed(allDiags)
-      log(LogLevel.Debug, "All AltParser alternatives failed.")
-    }
+      Assert.invariant(i <= limit)
+      if (returnFlag == false) {
+        Assert.invariant(i == limit)
+        // Out of alternatives. All of them failed.
+        val allDiags = new AltParseFailed(context.schemaFileLocation, pstate, diagnostics.reverse)
+        pstate.setFailed(allDiags)
+        log(LogLevel.Debug, "All AltParser alternatives failed.")
+      }
 
-    Assert.invariant(pBefore eq null)
-    pstate.popDiscriminator
+      pstate.popDiscriminator
+    } catch {
+      // Similar try/catch/finally logic for returning marks is also used in
+      // the RepAtMostTotalNParser and RepUnboundedParser. The logic isn't
+      // easily factored out so it is duplicated. Changes made here should also
+      // be made there. Only these parsers deal with taking marks, so this logic
+      // should not be needed elsewhere.
+      case t: Throwable => {
+        if (pBefore != null) {
+          markLeakCausedByException = true
+          if (!t.isInstanceOf[SchemaDefinitionDiagnosticBase] && !t.isInstanceOf[UnsuppressableException]) {
+            val stackTrace = new StringWriter()
+            t.printStackTrace(new PrintWriter(stackTrace))
+            Assert.invariantFailed("Exception thrown with mark not returned: " + t + "\nStackTrace:\n" + stackTrace)
+          }
+        }
+        throw t
+      }
+    } finally {
+      var markLeak = false;
+      if (pBefore != null) {
+        pstate.discard(pBefore)
```

Comment:
Ah, nevermind.  It looks like we don't always do a pBefore = null after a discard.

Reply:
It is not needed here just because this code is for cleaning up leaked marked (i.e. non-null values). Since nothing uses pBefore afterwards, setting to null isn't needed.